README.md
[![Pypi-version](https://img.shields.io/pypi/v/gmail-connector)](https://pypi.org/project/gmail-connector)
[![Python](https://img.shields.io/badge/python-3.8%20%7C%203.9%20%7C%203.10-blue)](https://www.python.org/)

[![pages-build-deployment](https://github.com/thevickypedia/gmail-connector/actions/workflows/pages/pages-build-deployment/badge.svg)](https://github.com/thevickypedia/gmail-connector/actions/workflows/pages/pages-build-deployment)
[![pypi-publish](https://github.com/thevickypedia/gmail-connector/actions/workflows/python-publish.yml/badge.svg)](https://github.com/thevickypedia/gmail-connector/actions/workflows/python-publish.yml)

[![Pypi-format](https://img.shields.io/pypi/format/gmail-connector)](https://pypi.org/project/gmail-connector/#files)
[![Pypi-status](https://img.shields.io/pypi/status/gmail-connector)](https://pypi.org/project/gmail-connector)

[![GitHub Repo created](https://img.shields.io/date/1599432310)](https://api.github.com/repos/thevickypedia/gmail-connector)
[![GitHub commit activity](https://img.shields.io/github/commit-activity/y/thevickypedia/gmail-connector)](https://api.github.com/repos/thevickypedia/gmail-connector)
[![GitHub last commit](https://img.shields.io/github/last-commit/thevickypedia/gmail-connector)](https://api.github.com/repos/thevickypedia/gmail-connector)

# Gmail Connector
Python module to, send SMS, emails and read `unread` emails in `inbox` folder.

###### Modules and Protocols
- `email` - Format emails as `MIMEMultipart` object, read emails from `bytes` and `str` and decode headers.
- `smtplib` - `SMTP` Simple Mail Transfer Protocol to connect to `gmail` server, do `auth` and perform outgoing tasks.
- `imaplib` - `IMAP` Internet Message Access Protocol to access messages in an email mailbox.

## Usage
`pip install gmail-connector`

### Env Vars
Store a `.env` file with the args:
```bash
gmail_user = 'username@gmail.com',
gmail_pass = '<ACCOUNT_PASSWORD>'
```

[Send SMS](https://github.com/thevickypedia/gmail-connector/blob/master/gmailconnector/send_sms.py)
```python
from gmailconnector.send_sms import Messenger

response = Messenger(
    phone='+11234567890',
    message='Test SMS using gmail-connector'
).send_sms()
if response.ok:
    print(response.json())
```
<details>
<summary><strong>More on <a href="https://github.com/thevickypedia/gmail-connector/blob/master/gmailconnector/send_sms.py">Send SMS</a></strong></summary>

:warning: Gmail's SMS Gateway has a payload limit. So, it is recommended to break larger messages into multiple SMS.

###### Additional args:
- **subject:** Subject of the message. Defaults to `Message from GmailConnector`
- **carrier:** Use any of ``at&t``, ``t-mobile``, ``verizon``, ``boost``, ``cricket``, ``us-cellular``. Defaults to `t-mobile`.
- **sms_gateway:** SMS gateway of the carrier. Defaults to ``tmomail.net`` since the default carrier is ``t-mobile``.
- **delete_sent:** Boolean flag to delete the outbound email from SentItems. Defaults to ``True``

> Note: If known, using the `sms_gateway` will ensure proper delivery of the SMS.
</details>

[Send Email](https://github.com/thevickypedia/gmail-connector/blob/master/gmailconnector/send_email.py)
```python
from gmailconnector.send_email import SendEmail

response = SendEmail(
        recipient='another_username@gmail.com',
        subject='Howdy!'
    ).send_email()
if response.ok:
    print(response.json())
```
<details>
<summary><strong>More on <a href="https://github.com/thevickypedia/gmail-connector/blob/master/gmailconnector/send_email.py">Send Email</a></strong></summary>

###### Additional args:
- **body:** Body of the email. Defaults to blank.
- **attachment:** Filename that has to be attached.
- **filename:** Custom name for the attachment. Defaults to the attachment name itself.
- **sender:** Name that has to be used in the email.
- **cc:** Email address of the recipient to whom the email has to be CC'd.
- **bcc:** Email address of the recipient to whom the email has to be BCC'd.

> Note: To send email to more than one recipient, wrap `recipient`/`cc`/`bcc` in a list.
>
> `recipient=['username1@gmail.com', 'username2@gmail.com']`
</details>

[Read Email](https://github.com/thevickypedia/gmail-connector/blob/master/gmailconnector/read_email.py)
```python
from gmailconnector.read_email import ReadEmail

reader = ReadEmail(folder='"[Gmail]/All Mail"')  # Folder defaults to inbox
response = reader.instantiate(category='SMALLER 500')  # Search criteria defaults to UNSEEN
if response.ok:
    unread_emails = reader.read_email(response.body)
    for each_mail in list(unread_emails):
        print(each_mail)
else:
    print(response.body)
```

### Linting
`PreCommit` will ensure linting, and the doc creation are run on every commit.

**Requirement**
<br>
`pip install --no-cache --upgrade sphinx pre-commit recommonmark`

**Usage**
<br>
`pre-commit run --all-files`

### Change Log
**Requirement**
`pip install --no-cache --upgrade changelog-generator`

**Usage**
`changelog`

### Pypi Module
[https://pypi.org/project/gmail-connector/](https://pypi.org/project/gmail-connector/)

### Runbook
[https://thevickypedia.github.io/gmail-connector/](https://thevickypedia.github.io/gmail-connector/)

### Repository
[https://github.com/thevickypedia/gmail-connector](https://github.com/thevickypedia/gmail-connector)

## License & copyright

&copy; Vignesh Sivanandha Rao, Gmail Connector

Licensed under the [MIT License](https://github.com/thevickypedia/gmail-connector/blob/master/LICENSE)


CHANGELOG
Change Log
==========

0.6.0 (12/13/2022)
------------------
- Fix deleting sent sms from sent mail
- Block active process until completion
- Simplify responder.py

0.5.9 (12/08/2022)
------------------
- Add a flag to make smtp check optional
- Upgrade sphinx

0.5.8 (11/15/2022)
------------------
- Remove OS restriction to get MX records
- Remove unreliable port number access check
- Switch python publish to run on release
- Update README.md

0.5.7 (10/30/2022)
------------------
- Include package data to upload validators to pypi

0.5.6 (10/30/2022)
------------------
- Improve email validation
- Create a dedicated dir based module for it
- Update README.md and requirements

0.5.5 (10/29/2022)
------------------
- Feature improvements on send_email.py
- Add a feature to support multiple attachments in an email
- Create an arg to fail email if attachment fails
- Create a module to validate email using SMTP port
- Add usage examples in README.md

0.5.4 (10/21/2022)
------------------
- Instantiate the object before sending SMS or email
- Add an authenticate property
- Update README.md and docstrings

0.5.3 (10/08/2022)
------------------
- Add an option to format email body as HTML
- Fix a bug when calling thread to delete sent email after sms
- Fix a bug in sphinx doc creation
- Update docstrings

0.5.2 (06/08/2022)
------------------
- Remove verbose and override flags when loading .env

0.5.1 (02/21/2022)
------------------
- Update version compatibility in README.md
- Import local modules from __init__
- Add a script to build locally
- Remove docs from actions

0.5.0 (02/19/2022)
------------------
- Add a warning if count is called outside ReadEmail
- Strip string converted email info in the yielded dict

0.4.9 (02/19/2022)
------------------
- Read mails in all folders with different status
- Convert email receive time into local timezone
- Remove print statements and use yield instead
- Yield email information as a dictionary
- Upload to pypi when committed to master
- Update docs, README.md and CHANGELOG

0.4.8 (01/12/2022)
------------------
- Take a custom filename for the attachment
- Default to attachment name without the path
- Remove unnecessary variables

0.4.7 (01/09/2022)
------------------
- Check for `phone` and `recipient` in env var
- Take gmail username instead of email address

0.4.6 (01/08/2022)
------------------
- Load env vars from a .env file
- Reformat docstrings
- Fix sent item after sending an SMS

0.4.5 (01/01/2022)
------------------
- Add SMTP port number to send email

0.4.4 (12/30/2021)
------------------
- Add more information to payload restriction

0.4.3 (12/30/2021)
------------------
- Add payload limit for SMS

0.4.2 (12/22/2021)
------------------
- Remove endpoint validator due to high inaccuracies

0.4.1 (12/22/2021)
------------------
- Mark internal methods as private
- Include private methods in sphinx docs

0.4.0 (12/22/2021)
------------------
- Fix phone number digit validation

0.3.9 (12/20/2021)
------------------
- Create a separate method to validate input arguments
- Change arg phone_number to phone

0.3.8 (12/20/2021)
------------------
- Fix import issues with module vs sphinx

0.3.7 (12/20/2021)
------------------
- Specify carrier based sms-gateway to increase success rate
- Introduce carrier, sms_gateway and delete_sent as optional arguments
- Use a third-party email-validator to check endpoint before sending the email
- Remove redundant variables
- Add requirements.txt, update README.md and .gitignore

0.3.6 (11/10/2021)
------------------
- Return responses as a class object instead of a dictionary
- Delete messages after sending an SMS
- Do not remove docs directory if version is not bumped
- Generate CHANGELOG in reverse

0.3.5 (10/16/2021)
------------------
- Add project URLs and package requirements to pypi
- Add markdown support to sphinx autodocs
- Add a condition check for version upgrade
- Update docs and changelog

0.3.4 (08/11/2021)
------------------
- Add new lines to the message start to separate subject and body of the SMS
- Update sphinx documentation to 4.1.2

0.3.3 (08/04/2021)
------------------
- Fix incorrect HTTP return codes

0.3.2 (07/24/2021)
------------------
- Remove logger module.
- Add exception handlers for Messenger class.
- Update docs and CHANGELOG
- Bump version.

0.3.1 (07/22/2021)
------------------
- Return a dictionary element after sending an email/SMS.
- Add status code and description to return dict.
- Update docs and CHANGELOG
- Bump version.

0.3.0 (07/19/2021)
------------------
- Allow users to add multiple recipients while sending email.
- Add CC and BCC options.
- Check if attachment file is available before trying to attach.
- Wrap recipient, cc and bcc items in a single list before email kick off.
- Remove sender arg and default to the user login email address.
- Fix version number format.

0.2.9 (07/19/2021)
------------------
- Add logging
- Remove print statements
- Bump version

0.2.8 (07/19/2021)
------------------
- Bump version to support github action
- Auto upload to pypi

0.2.7 (07/19/2021)
------------------
- auto upload to pypi when tagged a release version

0.2.6 (07/19/2021)
------------------
- onboard docs.yml but only prints a statement

0.2.5 (07/19/2021)
------------------
- Add badges
- Update README.md and CHANGELOG
- Bump version

0.2.4 (07/18/2021)
------------------
- Onboard `pypi` module
- Add `setup.py`, `setup.cfg`, `__init__.py`, `CHANGELOG`
- Update README.md and docs
- Move files to `gmailconnector` support package

0.2.3 (07/18/2021)
------------------
- Increase page width and update README.md

0.2.2 (07/18/2021)
------------------
- Onboard send_sms.py and update docs

0.2.1 (07/17/2021)
------------------
- Onboard sphinx auto generated documentation

0.2.0 (07/17/2021)
------------------
- Refactor read_email.py and add send_email.py
- Add pre-commit for linting
- Update README.md

0.1.9 (06/28/2020)
------------------
- update README.md


gmailconnector/read_email.py
import os
from datetime import datetime, timedelta, timezone
from email import message_from_bytes
from email.header import decode_header, make_header
from imaplib import IMAP4_SSL

import pytz
from dotenv import load_dotenv

from .responder import Response

if os.path.isfile('.env'):
    load_dotenv(dotenv_path='.env')


class ReadEmail:
    """Initiates Emailer object to authenticate and yield the unread emails.

    >>> ReadEmail

    """

    FOLDERS = ['inbox', '"[Gmail]/Important"', '"[Gmail]/Starred"', '"[Gmail]/All Mail"', '"[Gmail]/Sent Mail"',
               '"[Gmail]/Drafts"', '"[Gmail]/Spam"', '"[Gmail]/Trash"']
    CATEGORIES = ['ALL', 'SEEN', 'UNSEEN']
    ADDITIONAL_CATEGORIES = ['SMALLER', 'TEXT', 'SINCE']
    LOCAL_TIMEZONE = datetime.now(timezone.utc).astimezone().tzinfo

    def __init__(self, gmail_user: str = os.environ.get('gmail_user') or os.environ.get('GMAIL_USER'),
                 gmail_pass: str = os.environ.get('gmail_pass') or os.environ.get('GMAIL_PASS'),
                 folder: str = 'inbox'):
        """Initiates all the necessary args.

        Args:
            gmail_user: Login email address.
            gmail_pass: Login password.
            folder: Folder where the unread emails have to be read from.

        References:
            https://imapclient.readthedocs.io/en/2.1.0/_modules/imapclient/imapclient.html#IMAPClient.xlist_folders

        See Also:
            Uses broad ``Exception`` clause to catch login errors, since the same is raised by ``imaplib``
        """
        if folder not in self.FOLDERS:
            raise ValueError(
                f"The chosen folder {folder} is not available.\n"
                f"Available Options: {', '.join(self.FOLDERS)}"
            )
        self.folder = folder
        self.mail = None
        if not all([gmail_user, gmail_pass]):
            raise ValueError(
                'Cannot proceed without the args or env vars: `gmail_user` and `gmail_pass`'
            )
        self.mail = IMAP4_SSL('imap.gmail.com')  # connects to gmail using imaplib
        # noinspection PyBroadException
        try:
            self.mail.login(user=gmail_user, password=gmail_pass)  # login to your gmail account using the env variables
            self.mail.list()  # list all the folders within your mailbox (like inbox, sent, drafts, etc)
            self.mail.select(folder)
        except Exception:
            self.mail = None
        self.gmail_user = gmail_user

    def instantiate(self, category: str = 'UNSEEN') -> Response:
        """Searches the number of emails for the category received and forms.

        Args:
            category: Search criteria to be performed.

        References:
            - Categories, ``SMALLER``, ``TEXT``, and ``SINCE`` take additional values.

        Examples:
            ``'SMALLER 500'``
            ``'TEXT "foo-bar"'``
            ``'SINCE 17-Feb-2022'`` [OR] ``'SINCE 1645300995.231752'``

        References:
            https://imapclient.readthedocs.io/en/2.1.0/api.html#imapclient.IMAPClient.search

        Returns:
            Response:
            A Response class containing number of email messages, return code and the encoded messages itself.
        """
        if not self.mail:
            return Response(dictionary={
                'ok': False,
                'status': 401,
                'body': 'BUMMER! Unable to read your emails.\n\nTroubleshooting Steps:\n'
                        '\t1. Make sure your username and password are correct.\n'
                        '\t2. Logon to https://myaccount.google.com/lesssecureapps and turn ON less secure apps.\n'
                        '\t3. If you have enabled 2 factor authentication, use thee App Password generated.'
            })

        if category not in self.CATEGORIES and not any(cat in category for cat in self.ADDITIONAL_CATEGORIES):
            raise ValueError(
                f"The chosen search criteria {category} is not available.\n"
                f"Available Options: {', '.join(self.CATEGORIES)}"
            )

        return_code, messages = self.mail.search(None, category)
        if return_code != 'OK':
            return Response(dictionary={
                'ok': False,
                'status': 404,
                'body': 'Unable to read emails.'
            })

        num = len(messages[0].split())
        if not num:
            return Response(dictionary={
                'ok': False,
                'status': 204,
                'body': f'No {category.lower()} emails found for {self.gmail_user} in {self.folder}',
                'count': num
            })

        if return_code == 'OK':
            return Response(dictionary={
                'ok': True,
                'status': 200,
                'body': messages,
                'count': num
            })

    def _get_info(self, response_part: tuple) -> dict:
        """Extracts sender, subject and time received from response part.

        Args:
            response_part: Encoded tuple of the response part in the email.

        Returns:
            dict:
            A dictionary of sender, subject and received time.
        """
        original_email = message_from_bytes(response_part[1])
        if received := original_email.get('Received'):
            date = received.split(';')[-1].strip()
        else:
            date = original_email.get('Date')
        if '(PDT)' in date:
            datetime_obj = datetime.strptime(date, "%a, %d %b %Y %H:%M:%S -0700 (PDT)")
        elif '(PST)' in date:
            datetime_obj = datetime.strptime(date, "%a, %d %b %Y %H:%M:%S -0800 (PST)")
        else:
            datetime_obj = datetime.now()
        sender = make_header(decode_header((original_email['From']).split(' <')[0]))
        sub = make_header(decode_header(original_email['Subject'])) \
            if original_email['Subject'] else None
        # Converts pacific time to local timezone as the default is pacific
        local_time = datetime_obj.replace(tzinfo=pytz.timezone('US/Pacific')).astimezone(tz=self.LOCAL_TIMEZONE)
        received_date = local_time.strftime("%Y-%m-%d")
        current_date_ = datetime.today().date()
        # Replaces current date with today or yesterday to make it simpler
        if received_date == str(current_date_):
            receive = local_time.strftime("Today, at %I:%M %p")
        elif received_date == str(current_date_ - timedelta(days=1)):
            receive = local_time.strftime("Yesterday, at %I:%M %p")
        else:
            receive = local_time.strftime("on %A, %B %d, at %I:%M %p")
        return {'sender': str(sender).strip(), 'subject': str(sub).strip(), 'datetime': receive.strip()}

    def read_email(self, messages: list or str) -> str:
        """Prints unread emails one by one after getting user confirmation.

        Args:
            Takes the encoded message list as an argument. This is the body of the ``instantiate`` method.

        Returns:
            Response:
            A custom response class with properties: ok, status and body to the user.
        """
        for nm in messages[0].split():
            dummy, data = self.mail.fetch(nm, '(RFC822)')
            for each_response in data:
                if isinstance(each_response, tuple):
                    yield self._get_info(response_part=each_response)
        else:  # Executes when there is no break statement in the loop above
            if self.mail:
                self.mail.close()
                self.mail.logout()


if __name__ == '__main__':
    reader = ReadEmail(folder='"[Gmail]/All Mail"')
    response = reader.instantiate(category='SMALLER 500')
    if response.ok:
        unread_emails = reader.read_email(response.body)
        for each_mail in list(unread_emails):
            print(each_mail)
    else:
        print(response.body)
